### PR TITLE
New AFL++ fuzzer.py

### PR DESCRIPTION
**Pull request by andreafioraldi**
This AFL++ fuzzer.py works fine. There is the support to build configurations, but the conf is harcoded ATM.
I added also a small fix that includes -pthread by default because I experienced some errors while linking C++ (libc++ uses pthread).

### CODE REVIEW

File: experiment/generate_report.py [hunk 1]
```diff
@@ -58,11 +58,10 @@ def get_arg_parser():
         '--benchmarks',
```
Comment (jonathanmetzman) [inline experiment/generate_report.py]: I think you might need to pull since this undoes a change I made recently.

File: fuzzers/utils.py [hunk 2]
```diff
@@ -61,7 +61,8 @@ def append_flags(env_var, additional_flags, env=None):
     '-pthread', '-Wl,--no-as-needed', '-Wl,-ldl', '-Wl,-lm',
     '-Wno-unused-command-line-argument'
 ]
-NO_SANITIZER_COMPAT_CXXFLAGS = ['-stdlib=libc++'] + NO_SANITIZER_COMPAT_CFLAGS
+NO_SANITIZER_COMPAT_CXXFLAGS = ['-stdlib=libc++', '-pthread'] + \
```
Comment (jonathanmetzman) [inline fuzzers/utils.py]: I don't want to make you spend more time debugging linking issues, but I'm curious why this change was necessary.
1. Why don't you use these variables?
2. Is the change to CXXFLAGS necessary given this change?
3. Why does the flag need to be added here? Is what we were doing before to append not sufficient?
4. Can you change this to
```
NO_SANITIZER_COMPAT_CXXFLAGS = (['-stdlib=libc++', '-pthread'] +
   NO_SANITIZER_COMPAT_CFLAGS)
```
We prefer using paren to backslash to break up lines in python.

File: fuzzers/aflplusplus/fuzzer.py [hunk 3]
```diff
@@ -44,25 +71,91 @@ def build():
...
+    cmplog_target_binary_directory = (
+        get_cmplog_build_directory(target_binary_directory))
+    target_binary_name = os.path.basename(target_binary)
+    cmplog_target_binary = os.path.join(cmplog_target_binary_directory,
+                                        target_binary_name)
+
     afl_fuzzer.prepare_fuzz_environment(input_corpus)
+    os.environ['AFL_PRELOAD'] = '/afl/libdislocator.so'
+
+    flags = ["-d"]  # FidgetyAFL is better when runnign alone
```
Comment (jonathanmetzman) [inline fuzzers/aflplusplus/fuzzer.py]: nit: typo in "running".
Also I'm not sure what the comment means. Does it mean that when starting from no seed fidgety AFL is better (I think this is the case).

File: docker/base-builder/Dockerfile [hunk 4]
```diff
@@ -52,4 +52,4 @@ COPY --from=base-clang /usr/local/lib/libc++*.a /usr/local/lib/
 COPY --from=base-clang /usr/local/include/ /usr/local/include
 
 # Use libc++ thoughout the build.
-ENV CXXFLAGS -stdlib=libc++
\ No newline at end of file
+ENV CXXFLAGS -stdlib=libc++ -pthread
```
Comment (jonathanmetzman) [inline docker/base-builder/Dockerfile]: do you remember where this failed?
If not, it's fine I can look into this..

File: experiment/generate_report.py (hunk 1 above)
Comment (jonathanmetzman) [reply experiment/generate_report.py]: Actually I think the formatter did this.
Sorry it wasn't applied before.

File: fuzzers/aflplusplus/fuzzer.py (hunk 3 above)
Comment (andreafioraldi) [reply fuzzers/aflplusplus/fuzzer.py]: I mean when running alone without other parallel instances of AFL++.

File: docker/base-builder/Dockerfile (hunk 4 above)
Comment (andreafioraldi) [reply docker/base-builder/Dockerfile]: I tested libpng and afl-clang-fast++ failed to link the harness binary due to the lack of -pthread.
The printed errors were related to functions in libc++ that use pthread.

**Comment by andreafioraldi** [comment]
Yeah sorry for the late reply but I was at dinner.

**Comment by jonathanmetzman** [comment]
> Yeah sorry for the late reply but I was at dinner.

Don't worry, I landed because I assumed we weren't in the same timezone, so I wasn't expecting a response until tommrrow.

**Comment by andreafioraldi** [comment]
Btw you should use -d also for all other AFL based fuzzers. I noticed that you are performing the deterministic stages, this explains why MOPT results to be the best fuzzer. Deterministic stages are useful, in my experience, when having parallel instances. In AFL++ with didn't change the default behavior for retro compatibility with AFL, but in most cases it performs much better with -d, especially on structured formats like video formats.

File: fuzzers/aflplusplus/fuzzer.py (hunk 3 above)
Comment (jonathanmetzman) [reply fuzzers/aflplusplus/fuzzer.py]: Ah got it!

**Comment by jonathanmetzman** [comment]
> Btw you should use -d also for all other AFL based fuzzers. I noticed that you are performing the deterministic stages, this explains why MOPT results to be the best fuzzer. Deterministic stages are useful, in my experience, when having parallel instances. In AFL++ with didn't change the default behavior for retro compatibility with AFL, but in most cases it performs much better with -d, especially on structured formats like video formats.

Interesting! I have to think about doing this because I sort of prefer using the defaults for each fuzzer. I'm fine with individual fuzzers turning on knobs but I'm less sure about doing it for all AFL fuzzers.

**Comment by andreafioraldi** [comment]
AFLSmart for sure will benefit, @thuanpv will agree, they use this configuration by default (not here, probably he didn't notice that -d is missing).

**Comment by thuanpv** [comment]
I support the observation of @andreafioraldi. The -d option seems to work well. It was the reason why in the evaluation of our AFLSmart paper, we used that option for both two selected baselines, AFL and AFLFast, to have a fair comparison. Please see the 3rd paragraph in subsection 5.5 of our TSE'19 paper ((https://thuanpv.github.io/publications/TSE19_aflsmart.pdf) for more information. AFLSmart algorithm is not implemented for the deterministic stage yet so -d is its default option. I think having structure-aware mutations in that stage could be beneficial but there are some technical issues there. For example, since most of our new mutators (e.g., chunk deletion, chunk duplication, chunk swapping) could change the size of the seeds, how do we keep the stage deterministic along with the normal byte/bit-level mutation operators (e.g., bit flipping). It could also be the main reason why "splicing" was not implemented as part of the deterministic stage in AFL.
